File: src/main/java/javabot/operations/urlcontent/URLContentAnalyzer.java
package javabot.operations.urlcontent;

import com.google.inject.Singleton;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.stream.Stream;

@Singleton
public class URLContentAnalyzer {
    public final static String[] patterns = {
            "astebin",
            "mysticpaste.com",
            "pastie",
            "gist.github.com",
            "ideone.com",
            "docs.oracle.com.*api",
            "git.io",
    };
    public final static Map<String, Pattern> matchingPatterns = new HashMap<>();

    static {
        Stream.of(patterns).forEach(s ->
                matchingPatterns.put(s, Pattern.compile(".*" + s + ".*"))
        );
    }

    public boolean precheck(String url) {
        try {
            checkForBlacklistedSites(url);
            return true;
        } catch(ContentException e) {
            return false;
        }
    }
    public boolean check(String url, String title) {
        try {
            checkNulls(url, title);
            checkTitleToURLRatio(url, title);
            checkForBlacklistedSites(url);
            return true;
        } catch (ContentException e) {
            // log failure?
            return false;
        }
    }

    private void checkForBlacklistedSites(String url) throws ContentException {
        for (String pattern : patterns) {
            if (matchingPatterns.get(pattern).matcher(url).matches()) {
                throw new ContentException(String.format("Rejected: blacklisted site %s", url));
            }
        }
    }

    private void checkNulls(String url, String title) throws ContentException {
        if (null == title || "".equals(title.trim())) {
            throw new ContentException(
                    String.format("Rejected: Title content for %s was empty",
                            url));
        }
    }

    private void checkTitleToURLRatio(String url, String title) throws ContentException {
        // now break down the words in the title.
        String[] wordsInTitle = title.toLowerCase().split(" ");
        long words = Stream.of(wordsInTitle)
                .map(s -> s.replaceAll("[\\W]|_", ""))
                .filter(s -> s.length() > 2)
                .count();
        // generate a ratio of words to occurrences in the title.
        long hits = Stream.of(wordsInTitle)
                .map(s -> s.replaceAll("[\\W]|_", ""))
                .filter(s -> s.length() > 2)
                .filter(s -> url.toLowerCase().contains(s.toLowerCase()))
                .count();

        // ratio close enough to 1:1? reject.
        double ratio = (hits * 1.0) / words;
        if ((hits * 1.0) / words >= 0.8) {
            throw new ContentException(
                    String.format("Rejected: Ratio of content from  %s to %s was %f%n",
                            url,
                            title,
                            ratio));
        }
    }

    class ContentException extends Exception {
        public ContentException(String message) {
            super(message);
        }
    }
}

File: src/main/kotlin/javabot/operations/urlcontent/URLContentAnalyzer.kt
package javabot.operations.urlcontent

import jakarta.inject.Singleton
import java.util.HashMap
import java.util.Locale
import java.util.regex.Pattern
import java.util.stream.Stream

@Singleton
class URLContentAnalyzer {

    fun precheck(url: String): Boolean {
        try {
            checkForBlacklistedSites(url)
            return true
        } catch (e: ContentException) {
            return false
        }
    }

    fun check(url: String, title: String?): Boolean {
        try {
            checkNulls(url, title)
            checkTitleToURLRatio(url, title!!)
            checkForBlacklistedSites(url)
            return true
        } catch (e: ContentException) {
            // log failure?
            return false
        }
    }

    @Throws(ContentException::class)
    private fun checkForBlacklistedSites(url: String) {
        for (pattern in patterns) {
            if (matchingPatterns[pattern]?.matcher(url)?.matches() ?: false) {
                throw ContentException("Rejected: blacklisted site ${url}")
            }
        }
    }

    @Throws(ContentException::class)
    private fun checkNulls(url: String, title: String?) {
        if (null == title || "" == title.trim()) {
            throw ContentException("Rejected: Title content for ${url} was empty")
        }
    }

    @Throws(ContentException::class)
    private fun checkTitleToURLRatio(url: String, title: String) {
        // now break down the words in the title.
        val wordsInTitle = title.lowercase(Locale.getDefault()).split(" ")
        val words =
            wordsInTitle.map { s -> s.replace("[\\W]|_".toRegex(), "") }.count { s -> s.length > 2 }
        // generate a ratio of words to occurrences in the title.
        val hits =
            wordsInTitle
                .map { s -> s.replace("[\\W]|_".toRegex(), "") }
                .filter { s -> s.length > 2 }
                .count { s ->
                    url.lowercase(Locale.getDefault()).contains(s.lowercase(Locale.getDefault()))
                }
        // ratio close enough to 1:1? reject.
        val ratio = (hits * 1.0) / words
        if ((hits * 1.0) / words >= 0.8) {
            throw ContentException(
                "Rejected: Ratio of content from ${url} to ${title} was ${ratio}"
            )
        }
    }

    inner class ContentException(message: String) : Exception(message)

    companion object {
        val patterns: Array<String> =
            arrayOf(
                "astebin",
                "mysticpaste\\.com",
                "pastie",
                "gist\\.github\\.com",
                "ideone\\.com",
                "docs\\.oracle\\.com.*api",
                "git\\.io",
                "localhost",
                "127\\.0\\.0\\.1",
                "glot\\.io"
            )
        val matchingPatterns: MutableMap<String, Pattern> = HashMap()

        init {
            Stream.of(*patterns).forEach { s -> matchingPatterns.put(s, Pattern.compile(".*$s.*")) }
        }
    }
}
